torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -180,6 +181,43 @@ def _post_forward(
     return output
 
 
+@no_type_check
+def _fsdp_root_pre_forward(
+    state: _State,
+    *args,
+    **kwargs,
+):
+    """
+    Runs pre-forward logic specific to the root FSDP instance, which should run
+    before any individual module's pre-forward. If this is called on a non-root
+    FSDP instance, then the forward inputs are returned directly.
+    """
+    p_assert(state._is_root is not None, "Expects a root FSDP to have been set")
+    if not state._is_root:
+        return args, kwargs
+    if state.forward_prefetch:
+        handles_keys = []
+        if _is_composable(state):
+            # TODO: This assumes singleton handles keys.
+            handles_keys = [tuple(handle) for handle in state._handles]
+        else:
+            for fsdp_module in state.fsdp_modules(state):
+                handles_key = tuple(fsdp_module._handles)
```

Comment:
this seems not to put handles_key into handles_keys list?

Reply:
Great catch!